### ISSUE TITLE
deduplicate

### ISSUE THREAD

**Issue by allanwind** [newsboat/newsboat#1729]
**Newsboat version (copy the output of `newsboat -v` or the first line of `git show`)**:

Newsboat r2.24-70-g41f1b - https://newsboat.org/
Copyright (C) 2006-2015 Andreas Krennmair
Copyright (C) 2015-2021 Alexander Batischev
Copyright (C) 2006-2017 Newsbeuter contributors
Copyright (C) 2017-2021 Newsboat contributors

Newsboat is free software licensed under the MIT License. (Type `./newsboat -vv' to see the full text.)
It bundles:
- JSON for Modern C++ library, licensed under the MIT License: https://github.com/nlohmann/json
- optional-lite library, licensed under the Boost Software License: https://github.com/martinmoene/optional-lite
- expected-lite library, licensed under the Boost Software License: https://github.com/martinmoene/expected-lite

Newsboat r2.24-70-g41f1b
System: Linux 4.19.0-17-amd64 (x86_64)
Compiler: g++ 8.3.0
ncurses: ncurses 6.1.20181013 (compiled with 6.1)
libcurl: libcurl/7.64.0 OpenSSL/1.1.1d zlib/1.2.11 libidn2/2.0.5 libpsl/0.20.2 (+libidn2/2.0.5) libssh2/1.8.0 nghttp2/1.36.0 librtmp/2.3 (compiled with 7.64.0)
SQLite: 3.27.2 (compiled with 3.27.2)
libxml2: compiled with 2.9.4

Config file (copy from ~/.newsboat/config or ~/.config/newsboat/config):

```
article-sort-order title
articlelist-title-format "%U"
auto-reload yes
bind-key J next-feed
bind-key K prev-feed
bind-key j next
bind-key k prev
bind-key o open-in-browser-and-mark-read articlelist
datetime-format "%F %T"
error-log ~/.local/share/newsboat/error.log
feedlist-format "%4i %n %11u %L - %t"
# ignore-article "*" ...
ignore-mode display
keep-articles-days 365
prepopulate-query-feeds yes
reload-threads 32
run-on-startup end; open
show-read-articles no
show-read-feeds no
suppress-first-reload yes
unbind-key C
```

Steps to reproduce the issue:

1. I follow a couple of aggregation feeds: Hacker News, Lobsters and often times authors original feed.  This means I end up with 3 copies of some articles.  Added a `"query:Unread Articles:unread = \"yes\"` along with `article-sort-order title` to manually deduplicate those (i.e. mark two of those read in feedlist) but it would be better if newsboat could do this automatically probably using the Link attribute.  When a duplicate article is read or delete that action should happen on all 3 articles.  Not sure what the expect behavior should be when in the itemlist/itemview context (seeing duplicates as now, or take the same action across duplicates when operating on any of them).

**Other info you think is relevant**:

Low to medium priority.

**Comment by Minoru** [comment]
> I follow a couple of aggregation feeds: Hacker News, Lobsters and often times authors original feed.

Got it. Sounds quite similar to #898. When that bug is fixed, we could consider an option that treats the items's link as its GUID.

> I also subscribe to some forums where the Link values have a common prefix fora given thread:

The URLs are different here, so what should Newsboat deduplicate? Should it somehow devise that these two URLs belong to the same thread, and show only the first one?

**Comment by allanwind** [comment]
On 2021-07-31T10:15:48, Alexander Batischev wrote:
>> I follow a couple of aggregation feeds: Hacker News, Lobsters and often times authors original feed.
>
> Got it. Sounds quite similar to #898. When that bug is fixed, we could consider an option that treats the items's link as its GUID.

#898 wants duplicates (if I read it right), and #1729 is asking to 
hide them and take persistent action of all duplicates.

> The URLs are different here, so what should Newsboat 
> deduplicate? Should it somehow devise that these two URLs belong 
> to the same thread, and show only the first one?

I don't know if newsboat should thread conversions from forums,
in particular, as in if forum feed supply standardized information 
to group messages into threads and to order messages within a 
thread.  That said, if such an ID exist (perhaps rewriting the 
Link attribute with a regex rewrite), then ignoring an entire 
thread of messages is useful functionality, and would
be implemented similar to persistently deduplicate articles.  This 
probably should be a separate ticket now that I think about it,
but I meant it more of a heads-up as it's related.

**Comment by allanwind** [comment]
> The URLs are different here, so what should Newsboat 
> deduplicate? Should it somehow devise that these two URLs belong 
> to the same thread, and show only the first one?

I didn't answer your question, specifically, either flat or 
collapsed would work.  The latter UI choice carries semantic 
meaning so when you operate on a collapsed thread user wants to 
operate on a thread as a whole.

**Comment by Minoru** [comment]
Ah, right, the top post in #898 asks to keep duplicates separate. But the story of that bug goes way back, and one of the options is to keep it as a feature.

Regarding forums: if you find a somewhat-standard ID we can use, please open a separate issue and let's discuss how it could be supported in Newsboat.

**Comment by clappo143** [comment]
Hello, I am also looking to deduplicate items in Newsboat – specifically, automatically hide them from display or ignore/delete them in the first place. Any means of achieving this would be helpful. 

In my case, duplicate items (mostly news articles) will have identical Titles but do not necessarily come from the same source, meaning they will generally have entirely different URLs. 

I am almost entirely illiterate when it comes to this stuff, so apologies if the above does not make sense or would be better posted elsewhere. Any advice/suggestions or updates on this would be much appreciated. :)

**Comment by Minoru** [comment]
@clappo143 This is different from what we're discussing in this issue, so I moved your comment to #1891. Will post a comment there in a second.

**Comment by Friptick** [comment]
The feature as described in the title would be really useful. Often a news source will have multiple feeds based on the overlapping categories of its CMS ("Culture", "Politics" etc). So the same URLs will appear in different feeds. And if you use Newsboat's standard "All articles" view (`"query:Unread Articles:unread = \"yes\"`), you will end up with lots of annoying duplicates.

@Minoru: Seems the logical solution would be global options "deduplicate by URL" or "deduplicate by title". But this would need to work in the query views too.

@allanwind says:
>Added a "query:Unread Articles:unread = \"yes\" along with article-sort-order title to manually deduplicate those (i.e. mark two of those read in feedlist)

Not sure I understand, is this supposed to be a workaround? I have `article-sort-order date`, and in that same query view the duplicates are still there.

**Comment by allanwind** [comment]
Yes, an imperfect work-around, duplicates will often become runs
(index 7, 8, 9) then you can read one, and manually delete the
rest.  If you sort by date, then the duplicates will spread out
(index 1, 42, 201).

Don't forget about the same url across different sites
(where titles may not match).  lobste.rs vs 
https://news.ycombinator.com/
which has >50% overlap of content.

The other one that I hit regularly is feeds either screw up or 
move their content to a new blog (hotness du jour) and you are 
served all the articles again (duplicates in time).

Then there are "broken" feeds like Google News 
<https://news.google.com/atom?hl=en-US&gl=US&ceid=US%3Aen&oc=11> 
which is useless as a feed to me as it ends up showing me every
article on every sync (duplicates in time, broken feed).

**Comment by Friptick** [comment]
Thanks @allanwind.

>then you can read one, and manually delete the rest. 

Yep OK. Generous of you to call that a workaround!

>feeds either screw up or move their content to a new blog (hotness du jour) and you are served all the articles again

Yeah this is annoyingly common and gives RSS a bad name. And the suggested feature here would fix this too, in theory.

Thanks @Minoru for considering.

**Comment by Minoru** [comment]
@Friptick I think global options could be overkill, because you can't predict what URLs the feed authors will use. What if you enable an option because you wanted to deduplicate Lobsters and HN, but then another author posts something different linking to the same URL? I think these settings should operate on groups of feeds somehow, so you can explicitly say "I want deduplication between feeds A, B, and C".

I also just realized that this issue is very different from #898 in terms of UI: in #898, the duplicates are still displayed, it's just that operations like `toggle-unread` should affect either a single item or all of its copies. For this issue though, we want to hide duplicates entirely. The question, then, is *where* to hide them from if there are in multiple feeds. This leads me to think that the problem is better solved by a separate program that pulls the feeds, combines them into one, deduplicates them, and produces a new feed for Newsboat to consume. That would tie into Newsboat's `exec:` mechanism quite neatly, and also provide the grouping I described above. Please see https://newsboat.org/releases/2.28/docs/newsboat.html#_scripts_and_filters_snownews_extensions for pointers on how this could be implemented.

(This is not a hard NACK, so I'll keep this issue opened; but I now lean towards trying it out as an `exec:` script first, and if that works well and takes off, thinking of how to incorporate it into Newsboat proper.)

**Comment by Friptick** [comment]
@Minoru
>What if you enable an option because you wanted to deduplicate Lobsters and HN, but then another author posts something different linking to the same URL?

Yes, fair point. Those feeds contain crucial secondary URLs for each item.

>I think these settings should operate on groups of feeds somehow, so you can explicitly say "I want deduplication between feeds A, B, and C".

Could be a use for tags?

>This leads me to think that the problem is better solved by a separate program that pulls the feeds, combines them into one, deduplicates them, and produces a new feed for Newsboat to consume.

In fact I had already considered this solution but after reading and rereading the documentation about `execurl` AKA `exec`, and `filter`, I still feel in the dark about how exactly this would be done. I'm having trouble seeing if this would be a few lines of shell script or a giant XML-parsing monster. A boilerplate example would be great, deduplication seems like an obvious first feature to have somewhere.

**Comment by allanwind** [comment]
On 2022-08-07T13:30:56, Friptick wrote:
>In fact I had already considered this solution but after reading 
>and rereading the documentation about `execurl` AKA `exec`, and 
>`filter`, I still feel in the dark about how exactly this would 
>be done. I'm having trouble seeing if this would be a few lines 
>of shell script or a giant XML-parsing monster. A boilerplate 
>example would be great, deduplication seems like an obvious first 
>feature to have somewhere.

jq makes it relatively painless to work with json in a shell 
script.  There is xq (and others) for xml.

If you only want to deduplicate unread articles (~250/day for me) 
then would work better if it's passed all articles (link, title?) 
per exec, otherwise the script would have to maintain state.

If we want to support deduplication in time, then you either need 
to pass all articles (43k with keep-article-days 365), or perhaps 
better for newsboat to provide a suitable api to query the state 
that newsboat already maintains.

Besides deduplication we may also want to consider how we would
support group (tag) and sort order (both within a group and 
of all articles).

**Comment by Friptick** [comment]
This is looking more and more like a can of worms. One which I fear I do not have the expertise to help resolve. 

To take a step back and just make a couple of general remarks about UX.

At first glance, it seemed to me that since the query feed already "merges" separate feeds it should not be too hard to add a `deduplicate = \"yes\"` to just hide the dupes. But yeah, you also need to mark as read the hidden items. And manage the scenario of a already-seen URL reappearing tomorrow or next week, as @allanwind points out. But passing 43k articles to an external script seems somehow suboptimal.

>Besides deduplication we may also want to consider how we would
support group (tag) and sort order (both within a group and
of all articles).

Yes. In UX terms this might be an opportunity to remove the query feeds from `urls` entirely. The confusion of config with metadata in `urls` always felt a bit ugly to me. I see `urls` as a human-readable, editable, portable, state-maintaining list of feeds. It can be added to a backup sync regime. No other news reader allows this, they all ask you to remember to manually export a fussy OPML file. This is a little killer feature of Newsboat. But `urls` should contain generic feeds and generic tags, not strings of Newsboat-specific config vernacular. That should be in a config file. It may seem a nitpick but I can't resist sharing! @Minoru

Anyway, keep up the good work, it's appreciated.

**Comment by Minoru** [comment]
@Friptick Yeah, tags might be useful for this. Alternatively, we could just enumerate feed URLs as we do for `always-download` setting, but that then requires updates in two places when the URLs change.

The following examples might help with the script I proposed: https://github.com/newsboat/newsboat/blob/331176fdd9fdbd876f37b33f22a1a7054a13b2f0/contrib/slashdot.rb, https://github.com/newsboat/newsboat/blob/331176fdd9fdbd876f37b33f22a1a7054a13b2f0/contrib/heise.rb, https://github.com/msharov/snownews/blob/de3bd8b28191c4d4bc1be18275786613bcbc0c94/docs/untested/wikiwatch. They all fetch a single feed, change it, and print it out. My suggestion is to fetch multiple feeds, combine them, deduplicate the result, and print it out. I'd reach for Python with its `requests` and `feedparser` packages; shell is not really suitable for this IMHO.

urls file format is specific to Newsboat, so I don't see where the expectation of portability comes from. Moving query feeds out of urls file immediately raises the question of how to establish the ordering for `article-sort-order "none"`; I didn't think too deep on what other problems might come up. Please open a separate issue if you want to discuss this further.

@allanwind You seem to imagine a different design, one where the feed is piped through a script whenever it's opened in Newsboat. This doesn't exist, and I'm not sure it's wise to do, since feeds can be large; this will be slow. The `exec` I'm talking about is invoked much earlier, when the feed is downloaded by Newsboat.

**Comment by allanwind** [comment]
@AlexanderBatischev Actually, I did not have a design in mind, 
just reacting to the exec idea as I understood it and like you 
expecting performance issues given the volume of data involved.

If "fetch multiple feeds" means anything but all feeds, then it's
obviously a partial solution.  That is still better than what we 
have.  It would not address deduplicate from a previous fetch (I
see duplicates in every daily fetch of my feeds).  Solution 
implies a batch process where the ui would be effectively frozen 
for the 10+ seconds till all feeds have been downloaded.  It would 
be better to stream deduplicated articles.  This means that 
`itemlist` may update while user is looking at it.

**Comment by Friptick** [comment]
> The `exec` I'm talking about is invoked much earlier, when the feed is downloaded by Newsboat.

Exactly, but _how_ is it invoked earlier? This is the bit I'm finding hazy from the docs, maybe @allanwind is too. I mean, concretely, what is the `urls` file going to look like if certain feeds have been virtualized? An example would be great.

> What if you enable an option because you wanted to deduplicate Lobsters and HN, but then another author posts something different linking to the same URL?

@Minoru Back to the original request and this valid objection, how about a more minimal version of the feature, "Deduplicate items from same feed domain"? This would not affect independent aggregation site feeds, while addressing the common problem of a single site's "overlapping" feeds due to liberal tagging of articles. Just skip any item from a given feed's domain for which a matching URL already exists in the cache. Maybe update the existing cached item with "and x other feeds".

Just to clarify the use-case in case it still seems eccentric: if you subscribe to the NYT's feeds for Business, Culture, Politics, NY Regional News, Property, International, etc, you risk being overwhelmed with dupes. Such a shame because it undermines a killer feature of RSS compared to social media: the capacity to select topics from a single source with high granularity.

Anyway, I personally believe this feature would be entirely within the core scope of an RSS reader, since it's site-agnostic. You may still disagree and that's fine, it's your project after all. :)

> urls file format is specific to Newsboat, so I don't see where the expectation of portability comes from

Indeed, I meant data integrity more than portability. Which is even more important. For backup wouldn't it be progress to have a machine-readable, human-editable, always-up-to-date alternative to the drudgery of manually exporting OPML? Well, Newsboat almost has it already: the `urls` file is all of those things, and syncable. Any format is better than XML - YAML, TOML, CSV, space-separated values like Newsboat's `urls`, even maybe JSON. And personally I have repeatedly lost feed lists over the years because of the silly chore of having to remember OPML export. Maybe you weren't thinking about this issue, but Newsboat already solves it better than most, so well done.

**Comment by Minoru** [comment]
@allanwind 
> If "fetch multiple feeds" means anything but all feeds, then it's obviously a partial solution.

It does mean "not all feeds", and intentionally at that: as I described earlier, having deduplication applied to all feeds can lead to Newsboat dropping an item from an unrelated feed. So I want to scope it tight; the user has to somehow indicate a group of feeds for which items would be deduplicated. There could be multiple groups of course, or there might be a single group containing all feeds — whatever suits the user.

> It would not address deduplicate from a previous fetch (I see duplicates in every daily fetch of my feeds).

Ah, that's right, I didn't think of that. I still think it's worthwhile to experiment with an exec script before moving this functionality into Newsboat (at which point it will become retroactive and will affect already downloaded items).

@Friptick 
> I mean, concretely, what is the urls file going to look like if certain feeds have been virtualized? An example would be great.

Something like this:

```
"exec:~/.local/bin/group_and_dedup https://example.com/business.rss https://example.com/culture.rss https://example.com/politics.rss"
```

> if you subscribe to the NYT's feeds for Business, Culture, Politics, NY Regional News, Property, International, etc, you risk being overwhelmed with dupes

Hmm, looking at their source, there shouldn't be any risk; all items have GUIDs that don't seem to depend on the feed, i.e. the same item in two feeds will have the same GUID in both, and will be deduplicated by Newsboat. (Due to #898, the article will only appear in one of the feeds, and might migrate between them over time). Can you give an example of an article that appeared in two feeds at once? I'll need the URLs of the feeds and the title or URL of the article. If possible, give me the URLs upfront so I can subscribe and be ready when you tell me the article's title, otherwise I might miss it.

**Comment by Friptick** [comment]
Thanks @Minoru for the `exec` example, which completely demystifies the plugin mechanism! Maybe an idea to put it in the docs.

> Can you give an example of an article that appeared in two feeds at once?

In fact I do not read the NYT, I just assumed that the issue would affect any source which publishes lots of feeds with overlapping GUIDs. Here is a real example:

https://www.lemonde.fr/culture/rss_full.xml
https://www.lemonde.fr/cinema/rss_full.xml

If you subscribe to both these feeds, then the article today with `<guid isPermaLink="true">https://www.lemonde.fr/culture/article/2022/08/13/l-actrice-americaine-anne-heche-declaree-morte-une-semaine-apres-un-accident-de-voiture_6137910_3246.html</guid>`, which appears in both of them, will appear in the Newsboat query feed twice. I.e., duplicated.

>there shouldn't be any risk; all items have GUIDs that don't seem to depend on the feed, i.e. the same item in two feeds will have the same GUID in both, and will be deduplicated by Newsboat

OK now I'm confused. Are you saying that the behavior I just described is in fact a bug after all?

**Comment by allanwind** [comment]
I typically subscribe to an author's feed after seeing their
article references on an aggregating site.  This means my use
case would be to add every new feed to be deduplicated.  This
puts me in the opposite end of you in that I would much rather
miss an article than presented with duplicates.  This is just
point and not an argument for or against anything.

We may want to think of duplicates as a non-binary score perhaps
a k-means clustering of articles?  Then think about what
to do with a given cluster of similar articles.  If they are 
exact duplicates then it may not matter, but if they are similar
maybe user prefers feed A over B. I don't have any great
examples in my feed today but is an one example:

https://www.tomshardware.com/news/lowest-cpu-shipments-in-30-years-amd-intel-q2-2022-cpu-market-share
https://www.techspot.com/news/95611-desktop-cpu-sales-see-biggest-decline-30-years.html

The source of the information is the same (one of the articles, a 
3rd article, press release, embargo lifted of product release).  
In other words grouping and intra-group order including hiding
sufficiently similar articles.

Google News handles this in an interesting way on their website
(their feed is useless).  They have n references to the same
story, then elevate one of the sources as the main one and the
remaining n - 1 as secondary points (they still have duplicates
between sections for instance Headlines and U.S.).

**Comment by Minoru** [comment]
@Friptick The docs for `exec:` already have multiple examples, including one with a couple parameters. I honestly don't understand what's missing from there, and how adding an example with URLs instead of "param1" and "param2" would help. Can you maybe send a PR with that section reworked to your liking? Doesn't have to be perfect, just something to show what you expected to see there.

I missed the article you mentioned (my bad!), but I subscribed to both feeds and will keep an eye for a duplicate. Then I'll be able to tell if this is a bug or a feature request :)

**Comment by Friptick** [comment]
> I subscribed to both feeds and will keep an eye for a duplicate

Should have been plenty by now! :) I've been waiting to find out if this is a bug before investing an afternoon concocting a superfluous exec script. :)

Just to quickly recap if you're skimming this: the objective is to deduplicate items from different feeds where the item URL **and the item GUID** are both identical.

**Comment by Minoru** [comment]
Yep, I caught something a week ago, but haven't really looked into it 
yet. The GUIDs of the items are exactly the same, yet Newsboat treats 
them as separate items, which is confusing -- we have a bug where 
identical GUIDs should lead to the article "jumping" from feed to feed.

I'm busy participating in ICFP Programming Contest right now, but 
hopefully I'll find time next week to see what's going on with these 
feeds.

**Comment by Friptick** [comment]
Great! Thanks so much, and best of luck with the contest.

**Comment by Minoru** [comment]
…A-a-and now I'm down with COVID, so it's probably take at least one more week before I can dedicate any energy to this.

In case someone else wants to look into this, here are the feeds with duplicate items:  [cinema.rss.gz](https://github.com/newsboat/newsboat/files/9542927/cinema.rss.gz) [culture.rss.gz](https://github.com/newsboat/newsboat/files/9542929/culture.rss.gz) I expected the duplicate item to only show in one or the other (because of #898), but it somehow appears in both feeds within Newsboat, which is surprising. That's weird; I'd start with figuring out why this happens.

**Comment by Friptick** [comment]
Thanks for these initial thoughts @Minoru. But more importantly, take it easy and get well soon.

**Comment by Friptick** [comment]
To debug this, I have an idea for the easiest possible way to reproduce:

1. Subscribe to exactly the same feed twice! In `urls`, put the same URL on two lines.
2. Reload.

If the feed contains 10 items, my expectation is that I will see 10 items in the query feed view which contains both feeds. In practice, the query feed contains 10 duplicates.

Does this match your expectation?

**Comment by Minoru** [comment]
I think allowing the same URL multiple times is a bug :) Filed https://github.com/newsboat/newsboat/issues/2233 about that.

This scenario seems a bit different from what was discussed before. In this new scenario, there is only one feed in the database, and Newsboat just displays it twice. In the old scenario, there were two separate feeds that shared some of the articles. I understand why the new scenario behaves the way it does, but the old scenario is still a bit of a mystery to me -- and I haven't yet took the time to unravel it. Sorry!

**Comment by Friptick** [comment]
You're right, it's not exactly the same after all.

For unraveling purposes, those two Le Monde feeds are guaranteed to have overlap!

When you have time, of course. Thanks.

**Comment by Friptick** [comment]
@Minoru A quick question, tangentially related to this issue:

If an RSS feed contains an item with a URL _that has been marked as read_ but with a new date, is it expected behavior for the item to reappear in Newsboat as unread?

Example to be as clear as possible:

1. Day 1: feed contains item with title "XYZ", link "https://www.example.com/xyz.htm", date "14 November". I mark it as read.
2. Day 2: feed contains item with title "XYZ", link "https://www.example.com/xyz.htm", date "15 November". Boom! it appears in the feed again, unread.

Normal? I thought the URL would be the only UID used in the absence of a <guid> element. If not, would you consider adding a setting to make it so?

Use case: scraping a web page to generate the feed, when the publication dates are not available.

**Comment by Minoru** [comment]
@Friptick In the absence of a GUID Newsboat tries link+date, then just link, then the title, and finally sets GUID to empty: https://github.com/newsboat/newsboat/blob/3430491631bdb8c57e111c302e649e5f0a767b05/src/rssparser.cpp#L552 I really don't want to change this code because then its results will be different and people will suddenly get duplicate items in feeds that don't have GUIDs. Ditto with putting it behind a setting.

If you're generating the feed yourself, why not generate GUIDs too?

**Comment by der-lyse** [comment]
Looking at your dates, I were bitten by something similar at the weekend and it took me quite a bit to realize what was going on, so I might share, perhaps it helps somebody.

Besides other elements, my Atom feed's `<entry>` had `<id>`, `<link>` and `<updated>`, but no `<published>`. `<id>` and `<link>` were always fixed and did not change at any point in time. Before I updated my script to generate the feed, `<updated>` held the *creation timestamp* of the entry because of… well, I don't know. But after changing something in my script, `<updated>` suddenly contained the *last updated timestamp* of the entry. However, I was not aware of that accidental bugfix I somehow made.

In my config I have [`keep-article-days 365`](https://newsboat.org/releases/2.29/docs/newsboat.html#keep-articles-days). So this then caused old, read articles to be removed from the cache over time with the old script, because the creation timestamp was fixed. With the changed timestamp semantic in the `<updated>` fields, some of these timestamps then reflected some later point in time. Hence, new entries were recreated in Newsboat's cache, which of course are marked unread. That way, I ended up with unread, but old articles.

Newsboat only has one column for a timestamp in its cache. Just if there is no `<published>`, `<updated>` will be used as a fallback. Otherwise `<updated>` is not considered and ignored.

(I then fixed my script a second time to properly produce both `<published>` and `<updated>` with the correct timestamps.)

**Comment by Friptick** [comment]
> If you're generating the feed yourself, why not generate GUIDs too?

Because then I would need to maintain a database to link URLs to GUIDs. If you scrape the same date-free webpage each day, the only thing that will change is the URLs.

> In the absence of a GUID Newsboat tries link+date, then just link, then the title, and finally sets GUID to empty

Very useful, thanks. Seems the solution is to omit the `<pubdate>` for items where the scraper can't find one. Will try that.

> I really don't want to change this code

Of course. But don't forget the original (separate) issue in this ticket. Bump. :smile:

**Comment by Friptick** [comment]
> With the changed timestamp semantic in the `<updated>` fields, some of these timestamps then reflected some later point in time. Hence, new entries were recreated in Newsboat's cache, which of course are marked unread.

Thanks @der-lyse, this seems to confirm @Minoru's explanation.

**Comment by der-lyse** [comment]
No worries, @Friptick! I want to emphasize, key was the [`keep-article-days`](https://newsboat.org/releases/2.29/docs/newsboat.html#keep-articles-days) setting, though. If not specified, the cache is not cleaned up, so everything worked fine. With that setting, however, entries got dropped and then later were newly created. I could see that being reflected by a new database ID for the recreated entries. \</issue-hijacking> :-)

**Comment by Friptick** [comment]
@der-lyse Yes I see. Not exactly the same issue I'm having but definitely a gotcha if `keep-article-days` is lower than however many days it takes for a scraped webpage to be entirely refreshed. I'm going to adjust that straight away in anticipation, thanks. :+1:

**Comment by Minoru** [comment]
@Friptick You could use the URL itself as a GUID. That's what static site generators do, and it works well.

Haven't forgotten about the issue, it's just that answering your new question is much faster than digging into the issue ._.

**Comment by Friptick** [comment]
> You could use the URL itself as a GUID

Yep good idea, `<guid isPermaLink="true">URL</guid>`. Can't do any harm. BTW I just checked and that's what the "overlapping" feeds above do. I.e., two feeds, one GUID, item duplicated in query feed containing both. Maybe interesting.

>Haven't forgotten about the issue, it's just that answering your new question is much faster than digging into the issue ._.

Ha! No worries. Your project, you decide priorities. :+1:
